src/azul/service/repository_file_service.py

```diff
@@ -0,0 +1,132 @@
+        """
+        Return a file document from Elasticsearch.
+        """
+        filters = {
+            "fileId": {"is": [file_uuid]},
+            **({"fileVersion": {"is": [file_version]}}
+               if file_version else {})
+        }
+        es_search = self._create_request(catalog=catalog,
+                                         filters=filters,
+                                         post_filter=True,
+                                         enable_aggregation=False,
+                                         entity_type='files')
+        hits = [self._hit_to_doc(catalog, hit) for hit in es_search.scan()]
+        if not hits:
+            raise Exception('Unable to find file',
+                            catalog, file_uuid, file_version)
+        file = one([file for hit in hits for file in hit['contents']['files']])
+        if file_version:
+            assert file_version == file['version']
+        if file['drs_path'] is None:
+            raise Exception('File has no DRS path',
+                            catalog, file_uuid, file_version)
+        return file
+
+    def _get_access_url(self, drs_url: str, method_type: str) -> Mapping[str, str]:
+        """
+        Return the `access_url` dict for the specified `access_method` type from
+        the `drs_url` response.
+        """
+        tdr_client = TDRClient()
+        response = tdr_client.oauthed_http.request('GET', drs_url)
+        if response.status != 200:
+            raise RuntimeError('Failed to fetch file', response.data)
+        access_methods = json.loads(response.data)['access_methods']
+        access_method = one(am for am in access_methods if am['type'] == method_type)
+        return access_method['access_url']
+
+    def _get_blob(self, bucket_name, blob_name):
+        """
+        Get a blob object by name.
+        """
+        with shared_credentials():
+            client = gcs.Client()
+        bucket = gcs.Bucket(client, bucket_name)
+        return bucket.get_blob(blob_name)
+
+    def get_signed_url(self,
+                       catalog: CatalogName,
+                       file_uuid: str,
+                       file_version: Optional[str],
+                       expiration_seconds: int = 3600) -> str:
+        """
+        Return a signed url for the specified file.
+        """
+        file = self._get_file_document(catalog, file_uuid, file_version)
+        drs_uri = self.repository_plugin(catalog).drs_uri(file['drs_path'])
+        drs_client = DRSClient()
+        drs_url = drs_client.drs_uri_to_url(drs_uri)
+        access_url = self._get_access_url(drs_url, method_type='gs')
+        url_parts = furl(access_url['url'])
+        blob = self._get_blob(bucket_name=url_parts.netloc,
+                              blob_name='/'.join(url_parts.path.segments))
+        expiration = int(time() + expiration_seconds)
+        disposition = f"attachment; filename={file['name']}"
```

Comment:
There are some quoting rules that you need to be aware of. See https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition.

Those rules are more complex than what I can comprehend in the amount of time I'm willing to invest.

Do this

```py
file_name = '8" tape'
assert all(0x1f < ord(c) < 0x80 for c in file_name)
'attachment; filename="' + file_name.replace('"', r'\"') + '"'
'attachment; filename="8\\" tape"'
```

or equivalent.